firestore/src/main/java/com/firebase/ui/firestore/paging/FirestoreDataSource.java

```diff
@@ -216,18 +140,6 @@ public void onSuccess(QuerySnapshot snapshot) {
         @Override
         public void onFailure(@NonNull Exception e) {
             Log.w(TAG, "load:onFailure", e);
```

Comment:
So is there any way for the dev to recover the original error?

Reply:
Oops, seems like we have a small problem here because we're still using the DataSource class from Paging 2. I would like to bubble up this Exception to the adapter's LoadState listener, but I don't think DataSource provides a way to do that.

@samtstern any suggestions to better handle this error?

Reply:
Can we store the Exception and provide a `getLastFailure()` method or something?

Reply:
@samtstern Although `getLastFailure()` could be created on our `DataSource`, it would still not be exposed on the Adapter, because the Adapter has no reference to the DataSource. The DataSource object is created through a `DataSource.Factory` class.

With Paging 2, we were exposing the last error by getting the dataSource (`getDataSource()`) from a `PagedList`:

https://github.com/firebase/FirebaseUI-Android/blob/11dd3e7c4081b13984a75cd28d2375ebe7f5011f/firestore/src/main/java/com/firebase/ui/firestore/paging/FirestorePagingAdapter.java#L111-L118

But in Paging 3, `PagedList` has been replaced with `PagingData` and this new class does not provide us with a method to get the `DataSource`.

Reply:
I have a crazy idea in mind right now: Maybe we could replace our PagingSource with a [RemoteMediator](https://developer.android.com/topic/libraries/architecture/paging/v3-network-db#implement-remotemediator) since it handles network errors better than a PagingSource.

The RemoteMediator class was created to let developers load and sync _local_ and _network_ data easily. But maybe we could use it to load data from Firestore only.

I'll give this a try and share some feedback.